nodes/connect.go

```diff
@@ -66,23 +69,55 @@ func Connect(ctx context.Context, ns []p2plab.Node) error {
 		return err
 	}
 
-	connectPeers, gctx := errgroup.WithContext(ctx)
+	// Work out which addresses each node should dial, such that all dials
+	// are in one direction (nodes don't dial a node that dialed them).
+	// This helps avoid a libp2p bug where nodes that simultaneously dial
+	// each other can get random disconnects.
+	conns := make(map[peer.ID]map[peer.ID][]string)
+	for _, n := range ns {
+		npi, err := n.PeerInfo(ctx)
```

Comment:
`n.PeerInfo` is a network call that we previously made on line 48. I think we should make `peerInfos []peerstore.PeerInfo` to a `peerInfoByNodeID map[string]peerstore.PeerInfo` so we don't have to make the call again.